core/src/main/kotlin/org/evomaster/core/problem/api/service/ApiWsStructureMutator.kt

```diff
@@ -25,56 +26,118 @@ import kotlin.math.min
 /**
  * the abstract structure mutator for API based SUT, such as REST, GraphQL, RPC
  */
-abstract class ApiWsStructureMutator : StructureMutator(){
+abstract class ApiWsStructureMutator : StructureMutator() {
 
     companion object {
         private val log: Logger = LoggerFactory.getLogger(ApiWsStructureMutator::class.java)
     }
 
-
-    @Deprecated("External Actions will be moved into EnterpriseActionGroup")
-    private fun <T : ApiWsIndividual> addInitializingExternalServiceActions(
+    private fun <T : ApiWsIndividual> addExternalServiceActions(
         individual: EvaluatedIndividual<*>,
+        /**
+         * TODO add why
+         */
         mutatedGenes: MutatedGeneSpecification?,
+        /**
+         * [ExternalServiceHandler] is not injected in mutator at the moment, also to validate
+         * it's related to REST only sampler is needed
+         */
         sampler: ApiWsSampler<T>
     ) {
-        // TODO: Incomplete code, under development
 
         if (config.externalServiceIPSelectionStrategy == EMConfig.ExternalServiceIPSelectionStrategy.NONE) {
             return
         }
 
         if (sampler is ResourceSampler) {
-            val ind = individual.individual as? T
+            val ind = individual.individual as? ApiWsIndividual
                 ?: throw IllegalArgumentException("Invalid individual type")
 
-            // TODO: Under review
-            val fw = individual.fitness.getAccessedExternalServiceRequests()
+            val esr = individual.fitness.getViewAccessedExternalServiceRequests()
+            if (esr.isEmpty()) {
+                //nothing to do
+                return
+            }
 
-            val actions = mutableListOf<ExternalServiceAction>().plus(
-                sampler.getExternalService().getExternalServiceActions()
-            )
+            val newActions: MutableList<HttpExternalServiceAction> = mutableListOf()
 
-            ind.addInitializingDbActions(0, actions)
+            ind.seeMainExecutableActions().forEachIndexed { index, action ->
+                val parent = action.parent
+                if (parent !is EnterpriseActionGroup) {
+                    //TODO this should not really happen
+                    val msg = "Action is not inside an EnterpriseActionGroup"
+                    log.error(msg)
+                    throw RuntimeException(msg)
+                }
+
+                if (esr.containsKey(index)) {
+                    val urls = esr[index]
+
+                    if (urls!!.isNotEmpty()) {
+                        val actions = urls.map { url ->
+                            sampler.getExternalService()
+                                .getExternalServiceActionsForURL(url)
+                        }.flatten()
+
+                        if (actions.isNotEmpty()) {
```

Comment:
might `confirmeUsed` for all newly created actions

Reply:
if the URL is added in the `fv`, it indicates that it has been used, then we confirm it here.